akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -134,7 +134,7 @@ object ShardRegion {
         case ShardRegion.StartEntity(id) ⇒ id
         case _                           ⇒ entityId(message)
       }
-      (math.abs(id.hashCode) % maxNumberOfShards).toString
+      (math.abs(id.hashCode % maxNumberOfShards)).toString
```

Comment:
That's an interesting observation. For cluster sharding we can't change the end result of for example HashCodeMessageExtractor. That would be a disaster when doing a rolling upgrade. So I don't see how we could change this? What is the practical implication? It could be one more shard than specified (the size) and it will only contain the entities that hash to Int.MinValue?

Reply:
Good point. I believe that it is true for both `ShardingMessageExtractor` and `ShardRegion`. There can be a single additional shard created. It is still better than inconsistent shards when doing rolling updates as suggested. I think that these two changes can be reverted for now but they should be included in future when doing a major release and there are no rolling updates restrictions. Do you have a branch where changes for these two files can be committed?
What about all the other files? I think that in most cases we will end up accessing arrays with negative indexes if it is not fixed.